### DISCUSSION TITLE
Would a "redo-command" be useful?

### DISCUSSION THREAD

**Discussion by frosencrantz** [saulpw/visidata#2496] [Ideas]
I sometimes find that as I’m trying to add a column, split a value, etc. I end up in the loop of 1)  perform operation (e.g. `=` `addcol-expr`) and 2) enter input (e.g. python expression, regex, or shell command),  see that it did not work as expected, 3) hit the undo command, 4)  Rerun the command, 5) use the up arrow  and make edits.  I repeat the loop of step 3 to 5 until I get the result I want.

This involves a command that takes some input, and adds one or more columns.  I do not want the failed results as unwanted columns, and I want to edit the previous input.  This is why I might use a REPL elsewhere.

It is not hard to do this loop, but it would be nice if there was a `redo` command that would: 1) undo, 2) repeat the command that was undone, 3) populate the input with the previous input.

I’m not sure how hard this would be, and if this is a good solution. It would be useful if there was something that makes it possible to make this loop easier and faster.

**Comment by midichef** [comment]
Here's a quick prototype of this feature. I was surprised how much I like it. I didn't think I'd use it, but it makes fixing expressions pleasant and intuitive. So I vote that it should be a feature.

Try it out by adding this to your .visidatarc.
```
@VisiData.api
def revise_last_input(vd, sheet):
    # edited 2024-09-07 to handle inputMultiple() in an experimental way - use at your own risk!
    # edited 2024-08-31 to set the input 'type' parameter to the last type saved in input_history.jsonl
    # warning:  this edit may give incorrect results if a command is saved in command history but input is not saved to input history
    # or input is saved to input history but a command is not saved to command history
    vd.undo(sheet)
    sheet.undone or vd.fail("nothing to revise")
    cmdlogrow = sheet.undone[-1]
    if not cmdlogrow.input:
        vd.warning('last command had no input to revise')
    if isinstance(cmdlogrow.input, dict):
        kwargs = {}
# problem, we lose prompts and types for inputMultiple()
        for k, v in cmdlogrow.input.items():
            kwargs[k] = { 'type': None,
                          'prompt':'edit:  ',
                          'value': v
                        }
        cmdlogrow.input = vd.inputMultiple(**kwargs)
    else:
        lasttype = vd._inputHistoryList[-1]['type'] if vd._inputHistoryList else None
        cmdlogrow.input = vd.input(f'edit {cmdlogrow.longname} input:  ', value=cmdlogrow.input, type=lasttype)
    escaped = vd.replayOne(cmdlogrow)
    if not escaped:
        vd.status("%s redone" % cmdlogrow.longname)
        sheet.undone.pop()

BaseSheet.addCommand('zR', 'revise-last-input', 'vd.revise_last_input(sheet)', 'Revise input for the most recent command (options.undo must be enabled)')
```

I called it `revise-` instead of `redo-` to clarify that it differs from `redo-last` in that it does an undo.


Putting it in .visidatarc isn't a full implementation, because `revise-last-input` is left in the command log. To make it an actual feature, we'll need to exclude it from the cmdlog by putting it in `nonLogged`:
https://github.com/saulpw/visidata/blob/f5c07df6355fb8dc08179ba751f49302e02697ee/visidata/cmdlog.py#L12
which will become:
```
nonLogged = '''forget exec-longname undo redo revise-last-input quit
```

**Reply by saulpw** [reply]
@midichef nice implementation!

**Comment by frosencrantz** [comment]
Brilliant!

I quickly tried this and it works nicely, and I understand what you mean about being surprised by how much you like. I'll have to actually use it in a real loop to see if I have any feedback.

I am wondering about input history will that tie back to the input_history.jsonl file? It does not look like it, but I have not had time to really look.  Not sure if that will be possible.

Thank you @midichef!

**Reply by midichef** [reply]
As it is now, input into `revise-last-input` is saved in input_history.jsonl. But unfortunately it saves the input with a different `type` (`null`, in JSON) than the `type` used for many commands that have their own type, such as `"expr"` for `=`, `"regex"` for `/`, etc.

So looking back through the history of `=` input will show one set of values (with type `"expr"`). Looking back for `revise-last-input` input will go through a different set of values (with type `null`).

It's not currently possible to reliably find the`type` for input to the last command in the cmdlog within `.visidatarc`. That would require patching visidata itself.

**Reply by frosencrantz** [reply]
I was afraid of that.  I'm not opposed to those changes being made.  :). I was thinking your code here would be added to the feature tree.  I guess even without the typing information it is useful.

Maybe if the value was saved to the Memory Sheet it would be possible to paste into the input stream for the correct type.  The redo command is useful because I can quickly build a complicated expression, but then reuse it in a latter session, too.

This feels like it is a useful convenience.

**Reply by midichef** [reply]
Okay, thanks for explaining. Reusing expressions in later vd sessions would definitely be useful.

So I've updated the implementation for .visidatarc to try to guess what the last type was, using whatever was last in input history.

I don't think this is appropriate for a final implementation, because I suspect there are corner cases where vd would revise the wrong past input strings, because it is searching in the wrong input `type`. So I put a note warning about that. But if you only use `revise-last-input` for `=` it should work with no problems.

This should let you test the overall concept of this command further.

Let me know if you have any more feedback.

**Comment by frosencrantz** [comment]
I tried this with the `addcol-regex-subst` command, and it did not work as expected.  The input value was a json value like `{'before': 'pattern', 'after': 'replacement'}`.  Also, it will fail when I hit return with this value or a modified version of the json value.

It looks like input history is working. I have not tested this too closely yet.

I have not had to use this much yet, but still happy with the result.

**Reply by midichef** [reply]
Oh, thank you for the notice. It was not working for more than one input field, any time `inputMultiple()` was called in the last command.

I have just edited the snippet above to handle that situation. But `revise` cannot determine what the prompt should be then. This is why when using `revise-` on `addcol-regex-subst`, both prompts change from `replace:` and `search:` to `edit:` and `edit:`. That's the best I can do for a prototype in .visidatarc. To go any further, wiill require change to visidata internals, to make a record of input prompts.
